### PR TITLE
Implement sampling

### LINKED ISSUE #3: Transaction Sampling
Use the `sample_rate` configuration option to determine the percentage of sampled transactions. Note that non sampled transactions also do get reported. But they only contain the name and the duration, not the context and no spans.

### PR DESCRIPTION
**Pull request by felixbarny**
(closes #3)

Signed-off-by: Felix Barnsteiner <felix.barnsteiner@elastic.co>

### CODE REVIEW

File: apm-agent-core/src/main/java/co/elastic/apm/impl/sampling/ConstantSampler.java [hunk 1]
```diff
@@ -0,0 +1,29 @@
+package co.elastic.apm.impl.sampling;
+
+import co.elastic.apm.impl.transaction.TransactionId;
+
+public class ConstantSampler implements Sampler {
```
Comment (alvarolobato) [inline apm-agent-core/src/main/java/co/elastic/apm/impl/sampling/ConstantSampler.java]: Maybe some java doc saying what this is?

File: apm-agent-core/src/main/java/co/elastic/apm/impl/sampling/ProbabilitySampler.java [hunk 2]
```diff
@@ -0,0 +1,30 @@
+package co.elastic.apm.impl.sampling;
+
+import co.elastic.apm.impl.transaction.TransactionId;
+
+public class ProbabilitySampler implements Sampler {
```
Comment (alvarolobato) [inline apm-agent-core/src/main/java/co/elastic/apm/impl/sampling/ProbabilitySampler.java]: Maybe some java doc saying what this is?

File: apm-agent-core/src/main/java/co/elastic/apm/impl/sampling/Sampler.java [hunk 3]
```diff
@@ -0,0 +1,28 @@
+package co.elastic.apm.impl.sampling;
+
+import co.elastic.apm.impl.transaction.TransactionId;
+
+/**
+ * A sampler is responsible for determining whether a {@link co.elastic.apm.api.Transaction} should be sampled.
```
Comment (alvarolobato) [inline apm-agent-core/src/main/java/co/elastic/apm/impl/sampling/Sampler.java]: 👍

**Comment by alvarolobato** [comment]
BTW CI is failing, I guess it is they Keystore problem you are trying to fix in the other PR.

File: apm-agent-core/src/main/java/co/elastic/apm/impl/sampling/ConstantSampler.java (hunk 1 above)
Comment (felixbarny) [reply apm-agent-core/src/main/java/co/elastic/apm/impl/sampling/ConstantSampler.java]: done

File: apm-agent-core/src/main/java/co/elastic/apm/impl/sampling/ProbabilitySampler.java (hunk 2 above)
Comment (felixbarny) [reply apm-agent-core/src/main/java/co/elastic/apm/impl/sampling/ProbabilitySampler.java]: done

**Comment by felixbarny** [comment]
@Qard does `Transaction#isSampled()` sound strange to you as a native speaker?

**Comment by Qard** [comment]
No, seems reasonable to me.

**Pull request merged**